### ISSUE TITLE
Unknown binaries doesn't throw error

### ISSUE THREAD

**Issue by jaumard** [NodeOS/NodeOS#237]
When I use command that doesn't exist the command is simply ignored and I don't have a message to say something like `command not found`. I know this is a small things but I put here to the record ^^

**Comment by lite20** [comment]
Nsh, the command-line used in NodeOS is on the verge of being updated, but when I last checked a bug in NodeJS was stalling it's progress. I'm not sure if it's been resolved yet. @piranna will have more info on the status of it

**Comment by piranna** [comment]
> Nsh, the command-line used in NodeOS is on the verge of being updated, but when I last checked a bug in NodeJS was stalling it's progress. I'm not sure if it's been resolved yet. @piranna will have more info on the status of it

It seems it will be delayed a long time... I'm thinking about finding some workaround for this, but lately I don't have time for almost anything :-(

**Comment by luii** [comment]
@piranna where is the Bug i could look into it

**Comment by piranna** [comment]
> @piranna where is the Bug i could look into it

https://github.com/nodejs/node/issues/5574

**Comment by luii** [comment]
@piranna if i understand this correct you want to issue a command like nano and want to keep the process.stdin and duplicate it through the file descriptor for other commands.
This is a whole new level for me :D

**Comment by piranna** [comment]
It's never too late to start learning ;-)

**Comment by luii** [comment]
But is my assumption correct?

**Comment by piranna** [comment]
Yes, this is a low level bug on Node.js regarding management of tty
streams. I did it so easily on C some years ago, but this is Javascript.

**Comment by luii** [comment]
Maybe this is an alternative: [pty](https://github.com/chjj/pty.js)
Create and use a pseudo-tty via the pty module. This allows you to create a "fake" tty that allows you to run programs like `sudo` without actually hooking them up to a real tty. The benefit to this is that you can programmatically control/access stdin/stdout/stderr.

**Comment by piranna** [comment]
Yes, pty.js is an alternative... but it also has a bug where when using
process.stdin on pty.open() (we can't use pty.spawn() because we are not
using a single command), so it's not an option :-(

**Comment by piranna** [comment]
Anyway it's good to see your suggestion, it shows you are starting to think as an OS developer ;-)

**Comment by luii** [comment]
@piranna Would you mind to give me more information how nsh should work?
Why does it need tty? Does it really need multiple stdin's?
We could also use this information in the docs

**Comment by piranna** [comment]
> @piranna Would you mind to give me more information how nsh should work?

New `nsh` generate an AST tree from the bash source code and later walk and interpret it by creating `pipe()`able objects, also the external commands (as `Transform` streams). The tty is needed to create new `ReadStream` and `WriteStream` objects (the same ones used by the REPL)
